### ISSUE TITLE
[FEAT] Enable timestamp filtering from interval end towards interval start

### ISSUE THREAD

**Issue by ColinMoldenhauer** [sentinel-hub/eo-learn#498]
## **What is the problem? Please describe.**

For the training of a recurrent network, we require time-series data with a certain time difference between samples. Additionally, we require the final timestamp to be at a certain timestamp `t_ref`. Unfortunately, `SentinelHubInputTask` (or rather the underlying `sentinelhub.time_utils.filter_times()` method) queries the specified `time_interval` starting from the interval start asserting the specified `time_difference`. Therefore, it is almost impossible to find a suitable `time_interval`, which will produce an observation at `t_ref`.

## **Here's the solution**

The SentinelHubInputTask should support inverting the querying direction, i.e. starting with the end of `time_interval` backwards in time, respecting constraints of `time_difference`. To this end, a feature request has been filed with the `sentinelhub` package (see [Issue 347](https://github.com/sentinel-hub/sentinelhub-py/issues/347)) Once the feature has been added to `sentinelhub`, the changes to `eolearn` are sketched below:

### In `eolearn.io.sentinelhub_process.py`
- introduce `reverse_filter` field (see `__init__` stub)
- add `reverse_filter` parameter to `get_available_timestamps`
- change `SentinelHubInputTask._get_timestamp()` to allow reverse filtering
```python
class SentinelHubInputTask(SentinelHubInputBaseTask):
    def __init__(
        self,
        data_collection=None,
        size=None,
        resolution=None,
        bands_feature=None,
        bands=None,
        additional_data=None,
        evalscript=None,
        maxcc=None,
        time_difference=None,
        reverse_filter=False,
        cache_folder=None,
        max_threads=None,
        config=None,
        bands_dtype=None,
        single_scene=False,
        mosaicking_order=None,
        aux_request_args=None,
        session_loader: Optional[Callable[[], SentinelHubSession]] = None,
    ):
        ...
        self.reverse_filter = reverse_filter
        ...

    def _get_timestamp(self, time_interval, bbox):
        """Get the timestamp array needed as a parameter for downloading the images"""
        if self.single_scene:
            return [time_interval[0]]

        return get_available_timestamps(
            bbox=bbox,
            time_interval=time_interval,
            data_collection=self.data_collection,
            maxcc=self.maxcc,
            time_difference=self.time_difference,
            reverse_filter=self.reverse_filter,
            config=self.config,
        )

# copy of get_available_timestamps, where only filtering in the last line is replaced with 
# the (hopefully) upcoming new signature of sentinelhub.time_utils.filter_times
def get_available_timestamps(
    bbox: BBox,
    data_collection: DataCollection,
    *,
    time_interval: Optional[Tuple[dt.datetime, dt.datetime]] = None,
    time_difference: dt.timedelta = dt.timedelta(seconds=-1),
    reverse_filter: bool = False,
    maxcc: Optional[float] = None,
    config: Optional[SHConfig] = None,
) -> List[dt.datetime]:
    ...

    all_timestamps = search_iterator.get_timestamps()
    return filter_times(all_timestamps, time_difference, reverse_filter)

```

### In `sentinelhub.time_utils`
- added `reverse` parameter
- reverse the sorted timestamps if reverse=True
- use absolute value of timestamp difference
- reverse filtered timestamps again in order to recover order of filtered timestamps
```python
# an adapted version of filter_times, allowing for reverse filtering
def filter_times(timestamps: Iterable[TimeType],
                 time_difference: dt.timedelta,
                 reverse: bool = False) -> List[TimeType]:
    """Filters out timestamps within time_difference, preserving only...
        ... the oldest timestamp if reverse = False
        ... the most recent timestamp if reverse = True.

    :param timestamps: A list of timestamps.
    :param time_difference: A time difference threshold.
    :param reverse: Start filtering from the oldest (False) or most current (True) timestamp.
    :return: An ordered list of timestamps `d_1 <= d_2 <= ... <= d_n` such that `d_(i+1)-d_i > time_difference`.
    """
    timestamps = sorted(set(timestamps))
    if reverse: timestamps = timestamps[::-1]

    filtered_timestamps: List[TimeType] = []

    for current_timestamp in timestamps:
        if not filtered_timestamps or abs(current_timestamp - filtered_timestamps[-1]) > time_difference:
            filtered_timestamps.append(current_timestamp)

    if reverse: filtered_timestamps = filtered_timestamps[::-1]
    return filtered_timestamps

```

Best,
Colin

**Comment by zigaLuksic** [comment]
Hi @ColinMoldenhauer 

thanks for the suggestion. We'll try to see if there is a neat option for the user to provide the filtration function directly, to also solve any similar issues in the future.

**Comment by ColinMoldenhauer** [comment]
Good idea! Something along these lines?


#### In `eolearn.io.sentinelhub_process.py`
- introduce `filter_timestamps` parameter to `get_available_timestamps`
- introduce `filter_timestamps` parameter to `__init__()`
- save as "private" method
- pass filtering method in `_get_timestamp()`
```python
from sentinelhub import filter_times as sentinelhub_filter_times

class SentinelHubInputTask(SentinelHubInputBaseTask):
    def __init__(
        self,
        ....
        filter_timestamps=False,
        ...
    ):
		"""
                :param ...
		:param filter_timestamps: A Callable that filters the timestamps returned by get_available_timestamps, returning them in a list.
                : param ...
		"""
        ...
        self._filter_timestamps = filter_timestamps or sentinelhub_filter_times
        ...

    def _get_timestamp(self, time_interval, bbox):
        """Get the timestamp array needed as a parameter for downloading the images"""
        if self.single_scene:
            return [time_interval[0]]

        return get_available_timestamps(
            bbox=bbox,
            time_interval=time_interval,
            data_collection=self.data_collection,
            maxcc=self.maxcc,
            time_difference=self.time_difference,
            filter_timestamps=self._filter_timestamps,
            config=self.config,
        )

# copy of get_available_timestamps, where only filtering in the last line is replaced with 
# the newly passed function filter_timestamps
def get_available_timestamps(
    bbox: BBox,
    data_collection: DataCollection,
    *,
    time_interval: Optional[Tuple[dt.datetime, dt.datetime]] = None,
    time_difference: dt.timedelta = dt.timedelta(seconds=-1),
    filter_timestamps: Optional[Callable] = None,
    maxcc: Optional[float] = None,
    config: Optional[SHConfig] = None,
) -> List[dt.datetime]:
    ...
    filter_timestamps = filter_timestamps or sentinelhub_filter_times
    ....
    return filter_timestamps(all_timestamps, time_difference)

```

#### Define own filtering function somewhere, e.g. in `timestamp_filters.py`
```python
# an adapted version of filter_times, allowing for reverse filtering
def my_filter_times_reverse(timestamps: Iterable[TimeType],
                 time_difference: dt.timedelta) -> List[TimeType]:
    """Filters out timestamps within time_difference, preserving only the most recent timestamp.

    :param timestamps: A list of timestamps.
    :param time_difference: A time difference threshold.
    :return: An ordered list of timestamps `d_1 <= d_2 <= ... <= d_n` such that `d_(i+1)-d_i > time_difference`.
    """
    timestamps = sorted(set(timestamps))[::-1]

    filtered_timestamps: List[TimeType] = []

    for current_timestamp in timestamps:
        if not filtered_timestamps or filtered_timestamps[-1] - current_timestamp > time_difference:
            filtered_timestamps.append(current_timestamp)

    return filtered_timestamps[::-1]

```

#### Usage
```python
from timestamp_filters import my_filter_times_reverse

input_task = SentinelHubInputTask(
	...
	filter_timestamps=my_filter_times,
	...
)

```
Not sure if other tasks would need to be updated, too 🤷 

Best,
Colin

**Comment by zigaLuksic** [comment]
Something along these lines yes, anything else feels like we'd have to re-do it in a few months for another use-case.

We are planning a minor release this month (if all goes well), and this will likely be included. If you wish we can include you in the pull-request process once we get around to it.

**Comment by ColinMoldenhauer** [comment]
> We are planning a minor release this month (if all goes well), and this will likely be included. If you wish we can include you in the pull-request process once we get around to it.

That would be great!

**Comment by zigaLuksic** [comment]
was released in 1.3.1
